lib/active_merchant/billing/gateways/first_pay/first_pay_json.rb

```diff
@@ -0,0 +1,172 @@
+        post = {}
+        add_invoice(post, money, options)
+        add_reference(post, authorization)
+
+        commit(:refund, post)
+      end
+
+      def void(authorization, options = {})
+        post = {}
+        add_reference(post, authorization)
+
+        commit(:void, post)
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r(("processorId\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]').
+          gsub(%r(("merchantKey\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]').
+          gsub(%r(("cardNumber\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]').
+          gsub(%r(("cvv\\?"\s*:\s*\\?)[^,]*)i, '\1[FILTERED]')
+      end
+
+      private
+
+      def add_address(post, creditcard, options)
+        if address = options[:billing_address] || options[:address]
+          post[:ownerName] = address[:name]
+          post[:ownerStreet] = address[:address1]
+          post[:ownerCity] = address[:city]
+          post[:ownerState] = address[:state]
+          post[:ownerZip] = address[:zip]
+          post[:ownerCountry] = address[:country]
+        end
+      end
+
+      def add_invoice(post, money, options)
+        post[:transactionAmount] = amount(money)
+      end
+
+      def add_payment(post, payment, options)
+        post[:cardNumber] = payment.number
+        post[:cardExpMonth] = payment.month
+        post[:cardExpYear] = format(payment.year, :two_digits)
+        post[:cvv] = payment.verification_value
+      end
+
+      def add_reference(post, authorization)
+        post[:refNumber] = authorization
+      end
+
+      def commit(action, parameters)
+        response = parse(api_request(live_url + ACTIONS[action], post_data(parameters)))
+
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response),
+          error_code: error_code_from(response),
+          test: test?
+        )
+      end
+
+      def api_request(url, data)
+        ssl_post(url, data, headers)
+      rescue ResponseError => e
+        e.response.body
+      end
+
+      def parse(data)
+        JSON.parse data
+      end
+
+      def headers
+        { 'Content-Type' => 'application/json' }
+      end
+
+      def format_messages(messages)
+        messages.map { |message| message['message'] || message }.join(' ')
+      end
+
+      def success_from(response)
+        response['isSuccess']
+      end
+
+      def message_from(response)
+        return format_messages(response['errorMessages']) if response['errorMessages'].present?
+
+        return format_messages(response['validationFailures']) if response['validationFailures'].present?
+
+        response['data']['authResponse'] if response['data']['authResponse'].present?
+      end
+
+      def error_code_from(response)
+        return 'isError' if response['isError']
+
+        return 'validationHasFailed' if response['validationHasFailed']
+
+        nil
```

Comment:
💬 Comment:

There is no need for the return value here, by default a method will return nil without an explicit return value.